relayer/src/worker/packet.rs

```diff
@@ -115,34 +115,54 @@ fn handle_packet_cmd<ChainA: ChainHandle, ChainB: ChainHandle>(
     cmd: WorkerCmd,
 ) -> Result<(), TaskError<RunError>> {
     match cmd {
-        WorkerCmd::IbcEvents { batch } => handle_update_schedule(link, clear_interval, path, batch),
+        WorkerCmd::IbcEvents { batch } => {
+            handle_clear_packet(
+                link,
+                clear_interval,
+                path,
+                should_clear_on_start,
+                Some(batch.height),
+            )?;
+            handle_update_schedule(link, clear_interval, path, batch)
+        }
 
         // Handle the arrival of an event signaling that the
         // source chain has advanced to a new block.
-        WorkerCmd::NewBlock {
-            height,
-            new_block: _,
-        } => {
-            if *should_clear_on_start {
-                handle_clear_packet(link, clear_interval, path, Some(height))?;
-
-                // Clear the flag only if handle_clear_packet succeeds
-                *should_clear_on_start = false;
-                Ok(())
-            } else if should_clear_packets(clear_interval, height) {
-                handle_clear_packet(link, clear_interval, path, Some(height))
-            } else {
-                Ok(())
-            }
+        WorkerCmd::NewBlock { height, .. } => handle_clear_packet(
+            link,
+            clear_interval,
+            path,
+            should_clear_on_start,
+            Some(height),
+        ),
+
+        WorkerCmd::ClearPendingPackets => {
+            handle_clear_packet(link, clear_interval, path, should_clear_on_start, None)
         }
-
-        WorkerCmd::ClearPendingPackets => handle_clear_packet(link, clear_interval, path, None),
     }
 }
 
-/// Whether or not to clear pending packets at this `step` for the given height.
-/// Packets are cleared if `clear_interval` is not `0` and if we have reached the interval.
-fn should_clear_packets(clear_interval: u64, height: Height) -> bool {
+/// Whether or not to clear pending packets at this `step` for some height.
+/// If the relayer has been configured to clear packets on startup and that has not
+/// occurred yet, then packets are cleared.
+/// If a height is received, then packets are cleared if `clear_interval` is not
+/// `0` and if we have reached the interval.
+fn should_clear_packets(
+    clear_interval: u64,
+    should_clear_on_start: &mut bool,
+    height: Option<Height>,
+) -> bool {
+    if height.is_none() {
+        return false;
+    }
+
+    if *should_clear_on_start {
+        *should_clear_on_start = false;
+        return true;
+    }
+
+    let height = height.unwrap();
+
     clear_interval != 0 && height.revision_height % clear_interval == 0
 }
```

Comment:
In order to reduce duplicate checks of the `should_clear_on_start` flag in the `handle_packet_cmd` function, the boolean was moved into this function, but I'm not terribly keen on this refactor, as it bloats the logical complexity of what was originally a very simple check. I'm also not entirely sure that this logic is entirely sound; it was merely the cleanest way I came up with to implement this.